types/params.go

```diff
@@ -69,6 +69,10 @@ func DefaultBlockGossipParams() BlockGossipParams {
 
 // Validate validates the ConsensusParams to ensure all values
 // are within their allowed limits, and returns an error if they are not.
+// Expecting:
```

Comment:
why duplicate the code? is this what we want from documentation? step by step duplication of the code

Reply:
we should probably document `maxBlockSizeBytes` somewhere here https://tendermint.readthedocs.io/en/master/specification/block-structure.html

Reply:
I don't think this is duplicating the code, it saves folks from having to open the code and having to read each line, by line. That's what am trying to avoid. I had to go through the guts of it in order to write the tests themselves.

Reply:
> I had to go through the guts of it in order to write the tests themselves.

well, you are a developer, aren't you? you have to go through the guts sometimes

Reply:
Perhaps a better alternative is to include these conditions in the comments on the Params struct itself?

Reply:
Could you provide a similar example from Golang source code?

Reply:
> well, you are a developer, aren't you? you have to go through the guts sometimes
  
LOL I don't think this is fair in this case - these conditions are undocumented, so from the perspective of being able to test the functionality exposed to the user, it's fair to say this was to0 opaque - but for that reason we should put it on the Params struct rather than here on the Validate function

Reply:
`well, you are a developer, aren't you? you have to go through the guts sometimes`
I believe that the purpose of us making APIs is so make it easy for other people to build off our stuff easily. Imagine you had to read through all the source code of Go's APIs just to be able to use a function? My motivation here it is to try to improve developer experience/productivity and reduce the cognitive overload.

`Could you provide a similar example from Golang source code?`
I can find tonnes of examples:
- https://github.com/golang/go/blob/a6ae01a64a6a5ef17a4825cd486aa31f62481100/src/net/http/request.go#L1146-L1147
- https://github.com/golang/go/blob/a9afa4e933f3eff131f12e24bb0f5b9f3168ca14/src/os/dir.go#L7-L46
https://github.com/golang/go/blob/a9afa4e933f3eff131f12e24bb0f5b9f3168ca14/src/os/env.go#L46 
- https://github.com/golang/go/blob/a9afa4e933f3eff131f12e24bb0f5b9f3168ca14/src/os/exec/exec.go#L127-L139
- https://github.com/golang/go/blob/a6ae01a64a6a5ef17a4825cd486aa31f62481100/src/runtime/mksizeclasses.go#L171-L174
- https://github.com/golang/go/blob/a6ae01a64a6a5ef17a4825cd486aa31f62481100/src/io/io.go#L286-L294
- https://github.com/golang/go/blob/a6ae01a64a6a5ef17a4825cd486aa31f62481100/src/io/io.go#L296-L302
- https://github.com/golang/go/blob/a6ae01a64a6a5ef17a4825cd486aa31f62481100/src/io/io.go#L330-L336
- https://github.com/golang/go/blob/a6ae01a64a6a5ef17a4825cd486aa31f62481100/src/bytes/bytes.go#L115-L119
- https://github.com/golang/go/blob/a6ae01a64a6a5ef17a4825cd486aa31f62481100/src/math/big/rat.go#L463-L468

and many more.

@ebuchman `Perhaps a better alternative is to include these conditions in the comments on the Params struct itself?`
Perhaps, but I think it is very easy for code to go stale that way as it is detached from the actual validation code.

Reply:
oh, wow. sorry for being so harsh. admit that I was wrong. great examples! Agree that good API and dev experience/productivity are very important.

Reply:
Not a biggie @melekes :)